docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,201 @@
+
+The default value is to be determined pending analysis on any additional CPU overhead
+incurred by the `kubelet`.
+
+## Eviction Window
+
+The `kubelet` will evict pods from the node when the eviction threshold has been met
+for a specific window of time.  This allows the node to temporarily consume resources
+that go beyond the specified eviction window without inducing eviction unless that
+consumption was sustained.
+
+The `kubelet` will accept a flag to define the eviction window as follows:
+
+* `--eviction-window=:` - Duration of time where the node is able to exceed the
+specified eviction threshold before eviction occurs.
+
+The default value is to be determined pending analysis on the monitoring interval.
+
+## Eviction scenario
+
+Let's assume the operator started the `kubelet` with the following flags:
+
+* `--eviction-threshold=memory=256Mi`
+* `--eviction-window=30s`
+* `--eviction-monitoring-interval=10s`
+
+Then the `kubelet` will interact with `cAdvisor` every `10s` to introspect
+current node usage.
+
+At each monitoring interval, if a compute resource has reached it's eviction
+threshold, the `kubelet` will record an event related to the node describing
+which resource is low.
+
+If a compute resource has reached its eviction threshold over the `30s` period
+the `kubelet` will initiate eviction to attempt to reclaim the resource that
+has met its eviction threshold.  The `kubelet` will record an event when eviction
+is initiated on the node.
+
+## Eviction of Pods
+
+If the eviction criteria have been met, the `kubelet` will evict containers
+based on their quality of service guarantee.  Within a particular quality of
+service tier, the `kubelet` may use an eviction strategy to determine the
+candidate set of pods to evict.
+
+The `kubelet` will accept a flag to define the eviction strategy as follows:
+
+* `--eviction-strategy=:"` - The strategy to use when evicting pods
+when eviction criteria have been met.
+
+The `kubelet` will target `BestEffort`, `Burstable`, and `Guaranteed`
+pods in that order to find a candidate set of pods to fail to force reclamation
+of the starved compute resource.  Within each quality of service tier,
+the `kubelet` will evaluate each pod relative to the eviction strategy.
+
+If the `kubelet` fails a pod based on eviction criteria, an `Event` is created
+describing why the `Pod` was failed by the `kubelet`.
+
+This proposal describes the `greedy` strategy as the first candidate to
+implement based on the ability for operators to easily reason about it.
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
+starved compute resource relative to their scheduling guarantee.
+
+It ranks pods within a quality of service tier as follows:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
```

Comment:
If no pods in the `Burstable` class are exceeding their request, why will we have to evict any pods?
Is it because the node overhead became too high?
Eviction of pods with resource requests isn't ideal.